carma-messenger-core/j2735_convertor/src/psm_convertor.cpp

```diff
@@ -0,0 +1,271 @@
+{
+  out_msg.radius_of_curvature = in_msg.radius_of_curvature;
+  out_msg.confidence = in_msg.confidence;
+}
+
+void PSMConvertor::convert(const carma_v2x_msgs::msg::AccelerationSet4Way& in_msg, j2735_v2x_msgs::msg::AccelerationSet4Way& out_msg)
+{
+  // Convert Longitudinal and Lateral
+  out_msg.longitudinal = ValueConvertor::valueCavToJ2735<int16_t>(
+      in_msg.longitudinal, units::CM_PER_M, in_msg.presence_vector,
+      carma_v2x_msgs::msg::AccelerationSet4Way::ACCELERATION_AVAILABLE, j2735_v2x_msgs::msg::AccelerationSet4Way::ACCELERATION_UNAVAILABLE);
+
+  out_msg.lateral = ValueConvertor::valueCavToJ2735<int16_t>(in_msg.lateral, units::CM_PER_M, in_msg.presence_vector,
+                                                             carma_v2x_msgs::msg::AccelerationSet4Way::ACCELERATION_AVAILABLE,
+                                                             j2735_v2x_msgs::msg::AccelerationSet4Way::ACCELERATION_UNAVAILABLE);
+
+  // Convert Vertical
+  out_msg.vert =
+      ValueConvertor::valueCavToJ2735<int8_t>(in_msg.vert, units::FIFTIETH_G_PER_M_PER_SEC_SQR, in_msg.presence_vector,
+                                              carma_v2x_msgs::msg::AccelerationSet4Way::ACCELERATION_VERTICAL_AVAILABLE,
+                                              j2735_v2x_msgs::msg::AccelerationSet4Way::ACCELERATION_VERTICAL_UNAVAILABLE);
+
+  // Convert Yaw Rate
+  out_msg.yaw_rate = ValueConvertor::valueCavToJ2735<int16_t>(
+      in_msg.yaw_rate, units::CENTI_DEG_PER_DEG, in_msg.presence_vector,
+      carma_v2x_msgs::msg::AccelerationSet4Way::YAWRATE_AVAILABLE, j2735_v2x_msgs::msg::AccelerationSet4Way::YAWRATE_UNAVAILABLE);
+}
+
+void PSMConvertor::convert(const carma_v2x_msgs::msg::PositionalAccuracy& in_msg, j2735_v2x_msgs::msg::PositionalAccuracy& out_msg)
+{
+  // Convert semi_major Axis
+  out_msg.semi_major = ValueConvertor::valueCavToJ2735<uint8_t>(
+      in_msg.semi_major, units::TWENTIETH_M_PER_M, in_msg.presence_vector,
+      carma_v2x_msgs::msg::PositionalAccuracy::ACCURACY_AVAILABLE, j2735_v2x_msgs::msg::PositionalAccuracy::ACCURACY_UNAVAILABLE);
+
+  // Convert semi_minor Axis
+  out_msg.semi_minor = ValueConvertor::valueCavToJ2735<uint8_t>(
+      in_msg.semi_minor, units::TWENTIETH_M_PER_M, in_msg.presence_vector,
+      carma_v2x_msgs::msg::PositionalAccuracy::ACCURACY_AVAILABLE, j2735_v2x_msgs::msg::PositionalAccuracy::ACCURACY_UNAVAILABLE);
+
+  // Convert Orientation
+  out_msg.orientation = ValueConvertor::valueCavToJ2735<uint16_t>(
+      in_msg.orientation, units::DEG_360_OVER_65535_PER_DEG, in_msg.presence_vector,
+      carma_v2x_msgs::msg::PositionalAccuracy::ACCURACY_ORIENTATION_AVAILABLE,
+      j2735_v2x_msgs::msg::PositionalAccuracy::ACCURACY_ORIENTATION_UNAVAILABLE);
+}
+
+void PSMConvertor::convert(const carma_v2x_msgs::msg::PathHistory& in_msg, j2735_v2x_msgs::msg::PathHistory& out_msg)
+{
+    out_msg.curr_gnss_status = in_msg.curr_gnss_status;
```

Comment:
Before assigning, you should check presence vector for curr_gnss_status
if(in_msg.cur_gnss_status & carma_v2x_msg::msg::HAS_INITIAL_POSITION)

Same for initial position